File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/java/altova/text/edifact/DataCompletion.java
[call GenerateFileHeader("DataCompletion.java")]
package com.altova.text.edifact;

import com.altova.text.*;

import java.text.SimpleDateFormat;
import java.util.Date;

import javax.print.attribute.standard.DateTimeAtCompleted;

public abstract class DataCompletion {
	protected String m_StructureName = "";

	protected boolean hasKid(ITextNode node, String name) {
		return (0 < node.getChildren().filterByName(name).size());
	}

	protected ITextNode getKid(ITextNode node, String name) {
		return node.getChildren().filterByName(name).getAt(0);
	}

	protected void conservativeSetValue(ITextNode node, String value) {
		if (0 == node.getValue().length())
			node.setValue(value);
	}

	protected void conservativeSetValue(ITextNode node, char value) {
		if (0 == node.getValue().length()) {
			StringBuffer tmp = new StringBuffer();
			tmp.append(value);
			node.setValue(tmp.toString());
		}
	}

	protected void conservativeSetValue(ITextNode node, int value) {
		if (0 == node.getValue().length()) {
			StringBuffer tmp = new StringBuffer();
			tmp.append(value);
			node.setValue(tmp.toString());
		}
	}

	protected void conservativeSetValue(ITextNode node, long value) {
		if (0 == node.getValue().length()) {
			StringBuffer tmp = new StringBuffer();
			tmp.append(value);
			node.setValue(tmp.toString());
		}
	}

	protected String getStructureName() {
		return m_StructureName;
	}

	protected DataCompletion(String structurename) {
		m_StructureName = structurename;
	}

	public abstract void completeData(TextNode dataroot);

	protected String getCurrentDateAsEDIString() {
		Date now = new Date();
		SimpleDateFormat formatter = new SimpleDateFormat("yyMMdd");
		return formatter.format(now);
	}

	protected String getCurrentTimeAsEDIString() {
		Date now = new Date();
		SimpleDateFormat formatter = new SimpleDateFormat("HHmm");
		return formatter.format(now);
	}

	protected long getSegmentChildrenCount(ITextNode node) {
		long result = (ITextNode.Segment == node.getNodeClass()) ? 1 : 0;
		ITextNodeList children = node.getChildren();
		for (int i = 0; i < children.size(); ++i)
			result += getSegmentChildrenCount(children.getAt(i));
		return result;
	}
}

File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/cs/AltovaText/EDIFact/DataCompletion.cs
[call GenerateFileHeader("DataCompletion.cs")]
using System;

namespace Altova.TextParser.EDIFACT
{
	/// <summary>
	/// Encapsulates completing data.
	/// </summary>
	public abstract class DataCompletion
	{
		#region Implementation Detail:
		protected string mStructureName = "";
		#endregion
		#region Descendant interface:
		/// <summary>
		/// Constructs an instance of this class.
		/// </summary>
		/// <param name="structurename">the name of the structure to complete</param>
		protected DataCompletion(string structurename)
		{
			mStructureName = structurename;
		}
		/// <summary>
		/// Checks whether a node has a kid by a certain name.
		/// </summary>
		/// <param name="node">the node to check</param>
		/// <param name="name">the name to check for</param>
		/// <returns>true if the node has a kid named as specified, otherwise false</returns>
		protected static bool HasKid(ITextNode node, string name)
		{
			return (0 < node.Children.FilterByName(name).Length);
		}
		/// <summary>
		/// Returns a kid of a node by name.
		/// </summary>
		/// <param name="node">the node</param>
		/// <param name="name">the name of the kid to be returned</param>
		/// <returns>the named kid or (NEW!) NullNode if none was found</returns>
		protected static ITextNode GetKid(ITextNode node, string name)
		{
			ITextNode\[\]  kids = node.Children.FilterByName(name);
			if (kids.Length == 0)
				return NullTextNode.Instance;
			return kids\[0\];
		}
		/// <summary>
		/// Sets the value of a node if it is yet empty.
		/// </summary>
		/// <param name="node">the node whose value is to be set</param>
		/// <param name="value">the new value</param>
		protected static void ConservativeSetValue(ITextNode node, string value)
		{
			if (0 == node.Value.Length)
				node.Value = value;
		}
		/// <summary>
		/// Gets the name of the structure this instance will complete.
		/// </summary>
		protected string StructureName
		{
			get
			{
				return mStructureName;
			}
		}
		/// <summary>
		/// Formats the current data in a way suitable for EDI messages.
		/// </summary>
		/// <returns>the formatted date string</returns>
		protected static string GetCurrentTimeAsEDIString()
		{
			return DateTime.Now.ToString("HHmm");
		}
		/// <summary>
		/// Formats the current time in a way suitable for EDI messages.
		/// </summary>
		/// <returns>the formatted time string</returns>
		protected static string GetCurrentDateAsEDIString()
		{
			return DateTime.Now.ToString("yyMMdd");
		}
		/// <summary>
		/// Counts (recursively) all children of a node which are segment nodes.
		/// </summary>
		/// <param name="node">the node whose children to count</param>
		/// <returns>the number of children being segment nodes</returns>
		protected long GetSegmentChildrenCount(ITextNode node)
		{
			long result = (NodeClass.Segment == node.Class) ? 1 : 0;
			foreach (ITextNode kid in node.Children)
				result += GetSegmentChildrenCount(kid);
			return result;
		}
		#endregion
		#region Descendant obligations:
		/// <summary>
		/// Completes the hierarchy of nodes rooted in the passed node.
		/// </summary>
		/// <param name="dataroot">the root of the nodes to be completed</param>
		public abstract void CompleteData(ITextNode dataroot);
		#endregion
	}
}